designs/0003-ragged-containers.md

```diff
@@ -0,0 +1,575 @@
+may vary in size.
+
+### Ragged arrays of reals
+
+The simplest ragged container is a ragged array.  A simple non-trivial
+example is: 
+
+```
+real[{3, 4}] x = {{a, b, c}, {d, e, f, g}};
+```
+
+Here we are assuming that the standard array expression will be
+generalized to work for ragged arrays.  For example, `{1, 2}` is a
+two-dimensional array of integers, and `{{1, 2}, {3, 4}, {5, 6}}`
+denotes a three by two array of integers.
+
+The variable `x` is a ragged two-dimensional array of size two.
+Indexing works as usual, with
+
+```
+x[1] == {a, b, c}
+```
+
+and
+
+```
+x[2] == {d, e, f, g}
+```
+
+Sizing also works as expected, with
+
+```
+size(x) == 2
+```
+
+and
+
+```
+size(x[1]) == 3
+size(x[2]) == 4
+```
+
+This array is ragged because the sizes of `x[1]` and `x[2]` are
+different.  If `x[i1, ..., iN]` is the same for all valid indexes,
+then `x` is said to be *rectangular*.
+
+We write the unsized type of `x` the same way as for rectangular two-dimensional arrays,
+
+```
+x : real[ , ]
+```
+
+Declarations for ragged arrays are made up of the sizes of their
+elements.  Declaring `x` of type and size `real[{3, 4}]` declares `x`
+to be a two-dimensional array with `x[1]` having size three and `x[2]`
+size four.
+
+Providing an out-of-bounds index will raise the same exceptions (and
+thus cause the same behavior in interfaces) as other out-of-bounds
+errors.
+
+A ragged three-dimensional array may be defined similarly; each
+element in the sizing is the sizing of a ragged two-dimensional array.
+
+```
+real[{ {2, 3}, { 1, 2, 3 } }] y
+  = { {{a, b, c}, {d, e, f, g}},
+      {{h}, {i, j}, {k, l, m}} };
+```
+
+The size of `y` is the number of elements in its top-level
+declaration, here `size(y) == 2`.  The first element of `y` is a
+ragged two-dimensional array,
+
+```
+y[1] == {{a, b, c}, {d, e, f, g}}
+```
+
+and the second is
+
+```
+y[2] == {{h}, {i, j}, {k, l, m}}
+```
+
+The type of both `y[1]` and `y[2]` is `real[, ]`, the type of a
+real-valued two-dimensional array.  This is a requirement of ragged
+structures.
+
+This brings up an important principle of ragged structures:  each
+element in a ragged structure must be the same unsized type.  That is,
+it's not possible to have a ragged array whose first element is
+two dimensional and whose second element is three dimensional.
+
+### Ragged arrays of vectors
+
+In addition to arrays, the values of ragged structures may be vectors
+or matrices.  For example, a ragged array of vectors may be declared
+and defined as
+
+```
+vector[{3, 2}] v = { [a, b c]', [e, f]' };
+```
+
+The variable `v` is typed as an array of vectors, `v : vector[]`.  The
+value of `v[1]` is the size three vector `[a, b, c]'`, whereas the
+value of `v[2]` is the size two vector `[e, f]'`.  Both `v[1]` and
+`v[2]` are of type `vector`.  All 
+
+All size functions (`rows()`, `cols()`, etc.) work as expected.
+
+Row vectors are handled identically, e.g.,
+
+```
+row_vector[{3, 2}] w = { [a, b c], [e, f] };
```

Comment:
```suggestion
array[2] row_vector[{3, 2}] w = { [a, b c], [e, f] };
```
Again, the array here is optional. It does make it a bit more clear what is going on.